src/main/java/org/altbeacon/beacon/service/BeaconService.java

```diff
@@ -200,7 +202,14 @@ public void onCreate() {
                 BeaconManager.DEFAULT_FOREGROUND_BETWEEN_SCAN_PERIOD, mBackgroundFlag, mCycledLeScanCallback, bluetoothCrashResolver);
 
         beaconManager = BeaconManager.getInstanceForApplication(getApplicationContext());
-        beaconParsers = beaconManager.getBeaconParsers();
+
+        if (beaconManager.getBeaconParsers() != null) {
+            beaconParsers.addAll(beaconManager.getBeaconParsers());
```

Comment:
I made it a set to avoid duplicate `BeaconParser`s and optimize the advertisement parsing loop. This is done in case someone adds a `BeaconParser` as an extra data parser and also as a regular parser. But, you are right, adding a custom hashCode would also help.